src/hashsum/digest.rs

```diff
@@ -0,0 +1,266 @@
+    fn reset(&mut self);
+    fn output_bits(&self) -> usize;
+    fn output_bytes(&self) -> usize {
+        (self.output_bits() + 7) / 8
+    }
+    fn result_str(&mut self) -> String {
+        use self::rustc_serialize::hex::ToHex;
+
+        let mut buf: Vec<u8> = vec![0; self.output_bytes()];
+        self.result(&mut buf);
+        buf.to_hex()
+    }
+}
+
+impl Digest for md5::Context {
+    fn new() -> Self {
+        md5::Context::new()
+    }
+
+    fn input(&mut self, input: &[u8]) {
+        self.consume(input)
+    }
+
+    fn result(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(&*self.compute());
+    }
+
+    fn reset(&mut self) {
+        *self = md5::Context::new();
+    }
+
+    fn output_bits(&self) -> usize { 128 }
+}
+
+impl Digest for sha1::Sha1 {
+    fn new() -> Self {
+        sha1::Sha1::new()
+    }
+
+    fn input(&mut self, input: &[u8]) {
+        self.update(input);
+    }
+
+    fn result(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(&self.digest().bytes());
+    }
+
+    fn reset(&mut self) {
+        self.reset();
+    }
+
+    fn output_bits(&self) -> usize { 160 }
+}
+
+impl Digest for sha2::Sha224 {
+    fn new() -> Self {
+        sha2::Sha224::default()
+    }
+
+    fn input(&mut self, input: &[u8]) {
+        digest::Digest::input(self, input);
+    }
+
+    fn result(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(digest::Digest::result(*self).as_slice());
+    }
+
+    fn reset(&mut self) {
+        *self = sha2::Sha224::default();
+    }
+
+    fn output_bits(&self) -> usize { 224 }
+}
+
+impl Digest for sha2::Sha256 {
+    fn new() -> Self {
+        sha2::Sha256::default()
+    }
+
+    fn input(&mut self, input: &[u8]) {
+        digest::Digest::input(self, input);
+    }
+
+    fn result(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(digest::Digest::result(*self).as_slice());
+    }
+
+    fn reset(&mut self) {
+        *self = sha2::Sha256::default();
+    }
+
+    fn output_bits(&self) -> usize { 256 }
+}
+
+impl Digest for sha2::Sha384 {
+    fn new() -> Self {
+        sha2::Sha384::default()
+    }
+
+    fn input(&mut self, input: &[u8]) {
+        digest::Digest::input(self, input)
+    }
+
+    fn result(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(digest::Digest::result(*self).as_slice());
+    }
+
+    fn reset(&mut self) {
+        *self = sha2::Sha384::default();
+    }
+
+    fn output_bits(&self) -> usize { 384 }
+}
+
+impl Digest for sha2::Sha512 {
+    fn new() -> Self {
+        sha2::Sha512::default()
+    }
+
+    fn input(&mut self, input: &[u8]) {
+        digest::Digest::input(self, input)
+    }
+
+    fn result(&mut self, out: &mut [u8]) {
+        out.copy_from_slice(digest::Digest::result(*self).as_slice());
+    }
+
+    fn reset(&mut self) {
+        *self = sha2::Sha512::default();
+    }
+
+    fn output_bits(&self) -> usize { 512 }
+}
+
+impl Digest for sha3::Sha3_224 {
```

Comment:
I think it would be better to write a macro to generate most of these, as most of them are pretty similar to each other.